File: java/src/main/java/giis/selema/services/impl/VideoControllerLocal.java
package giis.selema.services.impl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import giis.portable.util.FileUtil;
import giis.selema.portable.selenium.CommandLine;
import giis.selema.services.IVideoController;

/**
 * Video controller to use when both the recorder container and the tests run in the same VM. Requires the container
 * running the test have access to docker and to the folders when the videos are stored. Any unexpected behaviour raises
 * an exception that must be handled by the caller video service.
 */
public class VideoControllerLocal implements IVideoController {
	final Logger log = LoggerFactory.getLogger(this.getClass());

	private String videoContainer;
	private String sourceFile;
	private String targetFolder;

	public VideoControllerLocal(String videoContainer, String sourceFile, String targetFolder) {
		this.videoContainer = videoContainer;
		this.sourceFile = sourceFile;
		this.targetFolder = targetFolder;
	}

	@Override
	public void start() {
		// First clean the video file (if exists) to prevent the recorder concatenating videos
		CommandLine.fileDelete(sourceFile, false);
		
		// The recorder should be created and stopped in order to start and record video now.
		// If not, stop now
		if (!"exited".equals(ContainerUtil.getContainerStatus(videoContainer))) {
			log.debug("Video recorder " + videoContainer + " is not stopped, restarting");
			ContainerUtil.runDocker("stop", videoContainer);
			ContainerUtil.waitDocker(videoContainer, "Shutdown complete", "", 5);
		}

		log.debug("Starting video recorder: " + videoContainer);
		ContainerUtil.runDocker("start", videoContainer);
		ContainerUtil.waitDocker(videoContainer, "Display", "is open", 5);
	}

	@Override
	public void stop(String videoName) {
		log.debug("Stopping video recorder: " + videoContainer);
		ContainerUtil.runDocker("stop", videoContainer);
		ContainerUtil.waitDocker(videoContainer, "Shutdown complete", "", 5);

		// copy the video file to its destination and then remove, this should not fail
		log.debug("Saving recorded video file to: " + videoName);
		CommandLine.fileCopy(sourceFile, FileUtil.getPath(targetFolder, videoName));
		CommandLine.fileDelete(sourceFile, true);
	}

}

File: video-controller/app/videoControllerLocal.js
import { CommandLine } from './commandLine.js';
import { ContainerUtil } from './containerUtil.js';
import { Log } from './log.js';

/**
 * Video controller for local Docker-based recording.
 * Assumes the test and recorder containers share the same VM and filesystem.
 * Any unexpected behavior throws an error to be handled by the caller.
 */
export class VideoControllerLocal {
h
  /**
   * Starts the video recording process by ensuring the container is clean and ready.
   * @returns {Promise<void>}
   */
  async start(videoContainer, sourceFile) {
    // Ensure container is stopped before starting
    const timestamp1 = Date.now();
    const status = await ContainerUtil.getContainerStatus(videoContainer);
    if (status !== 'exited') {
      Log.debug(`Video recorder ${videoContainer} is not stopped, restarting`);
      await ContainerUtil.runDocker('stop', videoContainer);
      await ContainerUtil.waitDocker(videoContainer, 'Shutdown complete', '', 5);
    }

    // Clean up previous video file to avoid concatenation
    await CommandLine.fileDelete(sourceFile, false);
    Log.trace("Time to ensure stopped: " + (Date.now() - timestamp1) + "ms");

    Log.debug(`Starting video recorder: ${videoContainer}`);
    const timestamp2 = Date.now();
    await ContainerUtil.runDocker('start', videoContainer);
    await ContainerUtil.waitDocker(videoContainer, 'Display', 'is open', 5);
    Log.trace("Time to ensure started: " + (Date.now() - timestamp2) + "ms");
  }

  /**
   * Stop the video recording process by ensuring the container is ready.
   * @returns {Promise<void>}
   */
  async stop(videoContainer) {
    Log.debug(`Stopping video recorder: ${videoContainer}`);
    const timestamp1 = Date.now();
    await ContainerUtil.runDocker('stop', videoContainer);
    await ContainerUtil.waitDocker(videoContainer, 'Shutdown complete', '', 5);
    Log.trace("Time to stop: " + (Date.now() - timestamp1) + "ms");
  }
}
